src/libworld/openflights.hpp

```diff
@@ -0,0 +1,420 @@
+//
+// This file is part of AT&C project which simulates virtual world of air traffic and ATC.
+// Code licensing terms are available at https://github.com/felix-b/atc/blob/master/LICENSE
+//
+#pragma once
+
+#include <fstream>
+#include <cstring>
+#include "libworld.h"
+
+using namespace std;
+using namespace world;
+
+// from https://stackoverflow.com/a/1120224
+class CSVRow
+{
+private:
+    string m_line;
+    vector<string> m_data;
+
+private:
+    // from https://stackoverflow.com/a/6089413
+    static istream &safeGetline(istream &is, string &t)
+    {
+        t.clear();
+
+        // The characters in the stream are read one-by-one using a std::streambuf.
+        // That is faster than reading them one-by-one using the std::istream.
+        // Code that uses streambuf this way must be guarded by a sentry object.
+        // The sentry object performs various tasks,
+        // such as thread synchronization and updating the stream state.
+
+        istream::sentry se(is, true);
+        streambuf *sb = is.rdbuf();
+
+        for (;;)
+        {
+            int c = sb->sbumpc();
+            switch (c)
+            {
+            case '\n':
+                return is;
+            case '\r':
+                if (sb->sgetc() == '\n')
+                    sb->sbumpc();
+                return is;
+            case streambuf::traits_type::eof():
+                // Also handle the case when the last line has no line ending
+                is.setstate(ios::eofbit);
+                if (t.empty())
+                {
+                    // No data, row is not valid : stream extraction has to fail
+                    is.setstate(ios::failbit);
+                }
+                return is;
+            default:
+                t += (char)c;
+            }
+        }
+    }
+
+public:
+    const string &operator[](size_t index) const
+    {
+        if (index >= size())
+            throw out_of_range{"CSVRow : index out of bounds"};
+        return m_data[index];
+    }
+    size_t size() const
+    {
+        return m_data.size();
+    }
+    void readNextRow(istream &str)
+    {
+        safeGetline(str, m_line);
+
+        m_data = split(m_line, ',');
+        // Add empty last field if needed
+        if (!m_line.empty() && (m_line.back() == ','))
```

Comment:
Sorry for not including the style guide - will try to do that today.
One important rule - let's always enclose the body of control statements (**if**/**for**/...) in `{` and `}`, even when the body is a one-liner.

Reply:
Got it.